server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java

```diff
@@ -0,0 +1,305 @@
+            AggregateCompletionStage<Void> setStage = CompletionStages.aggregateCompletionStage();
+            for (int i = 0; i < keyValuePairCount; i += 2) {
+               keyBytes = arguments.get(i);
+               byte[] valueBytes = arguments.get(i + 1);
+               setStage.dependsOn(cache.putAsync(keyBytes, valueBytes));
+            }
+            setStage.freeze().whenComplete((ignore, t) -> {
+               if (t != null) {
+                  log.trace("Exception encountered while performing MSET", t);
+                  ctx.writeAndFlush(stringToByteBuf("-ERR " + t.getMessage() + "\r\n", ctx.alloc()));
+               } else {
+                  ctx.writeAndFlush(stringToByteBuf("+OK\r\n", ctx.alloc()));
+               }
+            });
+            break;
+         case "INCR":
+            counterIncOrDec(cache, arguments.get(0), true)
+                  .thenAccept(longValue -> handleLongResult(ctx, longValue));
+            break;
+         case "DECR":
+            counterIncOrDec(cache, arguments.get(0), false)
+                  .thenAccept(longValue -> handleLongResult(ctx, longValue));
+            break;
+         case "INFO":
+            ctx.writeAndFlush(stringToByteBuf("-ERR not implemented yet" + "\r\n", ctx.alloc()));
+            break;
+         case "PUBLISH":
+            // TODO: should we return the # of subscribers on this node?
+            // We use expiration to remove the event values eventually while preventing them during high periods of
+            // updates
+            performSet(ctx, cache, SubscriberHandler.keyToChannel(arguments.get(0)),
+                  arguments.get(1), 3, type, ":0\r\n");
+            break;
+         case "SUBSCRIBE":
+            SubscriberHandler subscriberHandler = new SubscriberHandler();
+            return subscriberHandler.handleRequest(ctx, cache, type, arguments);
+         case "SELECT":
+            ctx.writeAndFlush(stringToByteBuf("-ERR Select not supported in cluster mode" + "\r\n", ctx.alloc()));
+            break;
+         case "READWRITE":
+         case "READONLY":
+            // We are always in read write allowing read from backups
+            ctx.writeAndFlush(stringToByteBuf("+OK\r\n", ctx.alloc()));
+            break;
+         case "RESET":
+            // TODO: do we need to reset anything in this case?
+            ctx.writeAndFlush(stringToByteBuf("+RESET\r\n", ctx.alloc()));
+         case "QUIT":
+            // TODO: need to close connection
+            ctx.flush();
+            break;
+         default:
+            return RequestHandler.super.handleRequest(ctx, cache, type, arguments);
+      }
+      return this;
+   }
+
+   private void handleLongResult(ChannelHandlerContext ctx, Long result) {
+      ctx.writeAndFlush(stringToByteBuf(":" + result + "\r\n", ctx.alloc()));
+   }
+
+   private CompletionStage<Long> counterIncOrDec(Cache<byte[], byte[]> cache, byte[] key, boolean increment) {
+      return cache.getAsync(key)
```

Comment:
why not use the clustered counters?

Reply:
At the time I wrote this it wasn't fully done. We can swap it out after this PR I would say.